Simulator/Utils/CreateFunctionMap.cpp

```diff
@@ -0,0 +1,151 @@
+/**
+ * @file CreateFunctionMap.cpp
+ *
+ * @ingroup Simulator/Utils
+ *
+ * @brief The templated CreateFunctionMap allows each class to create a concrete
+ * instance of itself at runtime.
+ *
+ */
+
+#include <map>
+#include <string>
+
+// Connections
+#include "Connections/NG911/Connections911.h"
+#include "Connections/Neuro/ConnGrowth.h"
+#include "Connections/Neuro/ConnStatic.h"
+
+// Edges
+#include "Edges/NG911/All911Edges.h"
+#include "Edges/Neuro/AllDSSynapses.h"
+#include "Edges/Neuro/AllDynamicSTDPSynapses.h"
+#include "Edges/Neuro/AllSTDPSynapses.h"
+#include "Edges/Neuro/AllSpikingSynapses.h"
+
+// Layout
+#include "Layouts/NG911/Layout911.h"
+#include "Layouts/Neuro/DynamicLayout.h"
+#include "Layouts/Neuro/FixedLayout.h"
+
+// Vertices
+#include "Vertices/NG911/All911Vertices.h"
+#include "Vertices/Neuro/AllIZHNeurons.h"
+#include "Vertices/Neuro/AllLIFNeurons.h"
+
+// Recorder
+#include "Recorders/IRecorder.h"
+#include "Recorders/NG911/Xml911Recorder.h"
+#include "Recorders/Neuro/XmlGrowthRecorder.h"
+#include "Recorders/Neuro/XmlSTDPRecorder.h"
+#include "Recorders/XmlRecorder.h"
+
+#if defined(HDF5)
+   #include "Recorders/Hdf5Recorder.h"
+   #include "Recorders/Neuro/Hdf5GrowthRecorder.h"
+#endif
+
+// MTRand
+#include "RNG/MTRand.h"
+#include "RNG/Norm.h"
+
+/**
+ *
+ * The templated CreateFunctionMap allows each class to create a concrete
+ * instance of itself at runtime. This approach helps to keep the code organized
+ * and avoids circular dependencies that can make the code less error-prone.
+ *
+ * How it avoids circular dependency?
```

Comment:
I think, at the start of this (or beforehand), you need to introduce that the factory needs to have a map that includes, for each concrete, derived class its name (string) and the creation function for that particular class. So, then the issue is "in what class do you populate this map, given that it needs to be done before the factory can instantiate any object?" That makes the circular dependency mentioned below clear. The downside (yeh, almost always is one) is that the approach chosen here requires that this factory template needs to include every definition of every class that might be instantiated. There is a plus, which is that all of the map entries are organized in this one file.

Reply:
Agreed, and changed accordingly.